### PR TITLE
FOUNDATIONS: Modify Ticket 

### PR DESCRIPTION
**Pull request by XusanDeveloper**

### CODE REVIEW

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+namespace Tarteeb.Api.Tests.Unit.Services.Foundations.Tickets
+{
+    public partial class TicketServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfTicketIsNullAndLogItAsync()
+        {
+            //given
+            Ticket nullTicket = null;
+            NullTicketException nullTicketException = new NullTicketException();
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs]: var

File: Tarteeb.Api/Services/Foundations/Tickets/TicketService.cs [hunk 2]
```diff
@@ -52,5 +53,17 @@ public partial class TicketService : ITicketService
 
             return maybeTicket;
         });
+
+        public ValueTask<Ticket> ModifyTicketAsync(Ticket ticket) =>
+        TryCatch(async () =>
+        {
+            ValidateTicketOnModify(ticket);
+            var maybeTicket = await this.storageBroker.SelectTicketByIdAsync(ticket.Id);
+
```
Comment (Javlonbek-dev) [inline Tarteeb.Api/Services/Foundations/Tickets/TicketService.cs:62]: Remove Line

File: Tarteeb.Api/Services/Foundations/Tickets/TicketService.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 // Free to use to bring order in your workplace
 //=================================
 
+using Microsoft.Extensions.Hosting;
 using System;
```
Comment (jamshidbek001) [inline Tarteeb.Api/Services/Foundations/Tickets/TicketService.cs:7]: order

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+                values: "Id is required");
+
+            var expectedTicketValidationException =
+                new TicketValidationException(invalidTicketException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime()).Returns(GetRandomDateTime);
+
+            //when
+            ValueTask<Ticket> addTicketTask = this.ticketService.ModifyTicketAsync(invalidTicket);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:108]: modifyTicketTask

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+            // then
+            actualTicketValidationException.Should().BeEquivalentTo(expectedTicketValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTicketByIdAsync(invalidTicket.Id),
+                    Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(),
+                    Times.Once);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:306]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+            this.loggingBrokerMock.Verify(broker =>
+               broker.LogError(It.Is(SameExceptionAs(
+                   expectedTicketValidationException))),
+                       Times.Once);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:311]: 3 line

**Review by jamshidbek001 — CHANGES_REQUESTED** [review]
please order using

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs (hunk 6 above)
Comment (XusanDeveloper) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs (hunk 5 above)
Comment (XusanDeveloper) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs (hunk 4 above)
Comment (XusanDeveloper) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:108]: Done

**Comment by ElbekDeveloper** [comment]
Conflicts

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+            this.storageBrokerMock.Verify(broker =>
+                broker.UpdateTicketAsync(It.IsAny<Ticket>()), Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsSameAsCreatedDateAndLogItAsync()
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:130]: UpdatedDateIsNotSameAs bo'lsa kerak

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+            //then
+            actualTicketValidationException.Should().BeEquivalentTo(
+                expectedTicketValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(expectedTicketValidationException))), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTicketByIdAsync(invalidTicket.Id), Times.Never);
+
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Tickets/TicketServiceTests.Validations.Modify.cs:163]: dateTimeBroker?

**Review by jamshidbek001 — COMMENTED** [review]
Add dateTimeBroker

**Review by jamshidbek001 — APPROVED** [review]

**Pull request merged**